File: solutions/java/0494.java
class Solution {
  public int findTargetSumWays(int[] nums, int S) {
    final int sum = Arrays.stream(nums).sum();

    if (sum < S || (sum + S) % 2 == 1)
      return 0;

    return subsetSum(nums, (sum + S) / 2);
  }

  private int subsetSum(int[] nums, int s) {
    int[] dp = new int[s + 1];
    dp[0] = 1;

    for (int num : nums)
      for (int w = s; w >= num; --w)
        dp[w] += dp[w - num];

    return dp[s];
  }
}

File: solutions/cpp/0494.cpp
class Solution {
 public:
  int findTargetSumWays(vector<int>& nums, int S) {
    const int sum = accumulate(begin(nums), end(nums), 0);
    if (sum < S || (sum + S) & 1) return 0;

    function<int(int)> subsetSum = [&](int target) {
      vector<int> dp(target + 1);
      dp[0] = 1;

      for (const int num : nums)
        for (int w = target; w >= num; --w)
          dp[w] += dp[w - num];

      return dp[target];
    };

    return subsetSum((sum + S) / 2);
  }
};